### PR TITLE
Add example of E2E testing

### PR DESCRIPTION
**Pull request by lelgenio**
## Motivation

Discussion: https://github.com/ipetkov/crane/issues/366

## Checklist

- [ ] added tests to verify new behavior
- [x] added an example template or updated an existing one
- [ ] updated `examples/README.md`

### CODE REVIEW

File: examples/end-to-end-testing/flake.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          ${workspace}/bin/server &
+          wait-for-connection --fail localhost:8000
+
+          # Firefox likes to write to $HOME
+          HOME="$(mktemp -d)"
+          geckodriver &
+          wait-for-connection localhost:4444
+
+          export SSL_CERT_FILE="${pkgs.cacert}/etc/ssl/certs/ca-bundle.crt"
+          ${workspace}/bin/e2e_tests
```
Comment (dpc) [inline examples/end-to-end-testing/flake.nix:62]: **thought (non-blocking):** Oh. Interesting. I think it works, and nothing wrong with it but generally [build a whole workspace for testing and other things, and then run the e2e script](https://github.com/fedimint/fedimint/blob/44faa9a9a1945914b1f6241d0f905a854235c81b/nix/craneBuild.nix#L140).

The advantage is that the test scripts work all the same both from dev env (`./script/run-some-test.sh`) and CI.

In Flakebox I even added specialized commands to make it cleaner.

Comment (lelgenio) [reply examples/end-to-end-testing/flake.nix:63]: Certainly! I added a comment to highlight this.

**Comment by lelgenio** [comment]
I went ahead and added PostgreSQL to the example, just to hammer the point home that you can spin up anything together in the test.

**Comment by lelgenio** [comment]
I didn't squash so it's easy to drop postgres if necessary

File: examples/end-to-end-testing/flake.nix [hunk 2]
```diff
@@ -28,10 +28,12 @@
           pname = "example-e2e";
           version = "0.1";
           doCheck = false;
-          nativeBuildInputs = lib.optionals pkgs.stdenv.isDarwin [
+          nativeBuildInputs = lib.optionals pkgs.stdenv.isDarwin (with pkgs.darwin.apple_sdk.frameworks; [
             pkgs.libiconv
-            pkgs.darwin.apple_sdk.frameworks.Security
-          ];
+            CoreFoundation
```
Comment (dpc) [inline examples/end-to-end-testing/flake.nix]: **question (non-blocking):** Do you know what are these for and what are a good defaults? AFAIR I always have to add `libiconv` and `Security`, but I don't even know why and what is the most commonly used set.

File: examples/end-to-end-testing/flake.lock [hunk 3]
```diff
@@ -0,0 +1,128 @@
+{
+  "nodes": {
```
Comment (ipetkov) [inline examples/end-to-end-testing/flake.lock:2]: We can omit this lockfile, none of the other examples do (I'm not entirely sure if Nix will blindly copy it when used as a template or not)

File: flake.nix [hunk 4]
```diff
@@ -82,6 +82,11 @@
           description = "Build a workspace with a trunk member";
           path = ./examples/trunk-workspace;
         };
+
+        end-to-end-testing = {
```
Comment (ipetkov) [inline flake.nix:86]: Nit: These are generally sorted alphabetically if you don't mind moving this up a tad!

File: examples/end-to-end-testing/flake.nix (hunk 2 above)
Comment (ipetkov) [reply examples/end-to-end-testing/flake.nix]: I think darwin builds usually complain if you don't have them (I always try without them and add them in if the build fails :man_shrugging: )

**Review by ipetkov — APPROVED** [review]
Looking great, thanks for adding this @lelgenio !

I don't mind having postgres in there as long as it doesn't end up making the test suite run for too long (though we can always comment it out later if it does).

Few minor nit/cleanup comments and happy to merge afterwards!

File: examples/end-to-end-testing/flake.lock (hunk 3 above)
Comment (lelgenio) [reply examples/end-to-end-testing/flake.lock:2]: Oops, how did that end up in here?

> I'm not entirely sure if Nix will blindly copy it when used as a template or not

Had to check, Yep the lockfile gets copied.

File: flake.nix (hunk 4 above)
Comment (lelgenio) [reply flake.nix]: Done. I let `default` stay together with `quick-start` though, as formatting seems to imply `default` is not an actual item on the list.

File: examples/end-to-end-testing/flake.nix (hunk 2 above)
Comment (lelgenio) [reply examples/end-to-end-testing/flake.nix]: I figured it out from the [logs](https://github.com/ipetkov/crane/actions/runs/6490015759/job/17625187508#step:5:2581), no idea what exactly it does.

image

**Pull request merged**